#Text Summarization with BERT

In [ ]:
!pip install transformers --quiet
!pip install bert-extractive-summarizer --quiet
!pip install spacy --quiet
!pip install torch --quiet

In [ ]:
from summarizer import Summarizer,TransformerSummarizer

In [ ]:
text = '''
Scientists say they have discovered a new species of orangutans on Indonesia’s island
of Sumatra.
The population differs in several ways from the two existing orangutan species found in
Sumatra and the neighboring island of Borneo.
The orangutans were found inside North Sumatra’s Batang Toru forest, the science publication
Current Biology reported.
Researchers named the new species the Tapanuli orangutan. They say the animals are considered
a new species because of genetic, skeletal and tooth differences.
Michael Kruetzen is a geneticist with the University of Zurich who has studied the orangutans
for several years. He said he was excited to be part of the unusual discovery of a new great
ape in the present day. He noted that most great apes are currently considered endangered or
severely endangered.
Gorillas, chimpanzees and bonobos also belong to the great ape species.
Orangutan – which means person of the forest in the Indonesian and Malay languages - is the
world’s biggest tree-living mammal. The orange-haired animals can move easily among the trees
because their arms are longer than their legs. They live more lonely lives than other great
apes, spending a lot of time sleeping and eating fruit in the forest.
The new study said fewer than 800 of the newly-described orangutans exist. Their low numbers
make the group the most endangered of all the great ape species.
They live within an area covering about 1,000 square kilometers. The population is considered
highly vulnerable. That is because the environment which they depend on is greatly threatened
by development.
Researchers say if steps are not taken quickly to reduce the current and future threats, the
new species could become extinct “within our lifetime.”
Research into the new species began in 2013, when an orangutan protection group in Sumatra
found an injured orangutan in an area far away from the other species. The adult male
orangutan had been beaten by local villagers and died of his injuries. The complete skull was
examined by researchers.
Among the physical differences of the new species are a notably smaller head and frizzier
hair. The Tapanuli orangutans also have a different diet and are found only in higher forest
areas.
There is no unified international system for recognizing new species. But to be considered,
discovery claims at least require publication in a major scientific publication.
Russell Mittermeier is head of the primate specialist group at the International Union for the
Conservation of Nature. He called the finding a “remarkable discovery.” He said it puts
responsibility on the Indonesian government to help the species survive.
Matthew Nowak is one of the writers of the study. He told the Associated Press that there are
three groups of the Tapanuli orangutans that are separated by non-protected land.He said
forest land needs to connect the separated groups.
In addition, the writers of the study are recommending that plans for a hydropower center in
the area be stopped by the government.
It also recommended that remaining forest in the Sumatran area where the orangutans live be
protected.
I’m Bryan Lynn.
'''

In [ ]:
bert_model = Summarizer()
bert_summary = ''.join(bert_model(text, min_length=60))
bert_summary

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.34G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


'Scientists say they have discovered a new species of orangutans on Indonesia’s island\nof Sumatra. They say the animals are considered\na new species because of genetic, skeletal and tooth differences. Orangutan – which means person of the forest in the Indonesian and Malay languages - is the\nworld’s biggest tree-living mammal. Their low numbers\nmake the group the most endangered of all the great ape species. He said it puts\nresponsibility on the Indonesian government to help the species survive. It also recommended that remaining forest in the Sumatran area where the orangutans live be\nprotected.'

#Text Summarization with GPT-2

In [ ]:
from summarizer import TransformerSummarizer

GPT2_model = TransformerSummarizer(transformer_type="GPT2",transformer_model_key="gpt2-medium")
GPT_summary = ''.join(GPT2_model(text, min_length=60))
GPT_summary

config.json:   0%|          | 0.00/718 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.52G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


'Scientists say they have discovered a new species of orangutans on Indonesia’s island\nof Sumatra. Their low numbers\nmake the group the most endangered of all the great ape species. They live within an area covering about 1,000 square kilometers. The Tapanuli orangutans also have a different diet and are found only in higher forest\nareas. Russell Mittermeier is head of the primate specialist group at the International Union for the\nConservation of Nature. He told the Associated Press that there are\nthree groups of the Tapanuli orangutans that are separated by non-protected land.'

#Text Summarization with XLNet

In [ ]:
!pip install keras-xlnet
!pip install sentencepiece

  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Created wheel for keras-xlnet: filename=keras_xlnet-0.20.0-py3-none-any.whl size=43564 sha256=b842b86ef8d5fb293bf198f9c6e9d66b3d3370fd29b1fed7dbb582acdf07d0dd
  Stored in directory: /root/.cache/pip/wheels/08/d5/78/676c2872b22be584434695a093ef3d560f2badade95b682aef
  Created wheel for keras-trans-mask: filename=keras_trans_mask-0.6.0-py3-none-any.whl size=3631 sha256=d61cce04f88cc956c01ceaf8d70e4e0692abf9f6cd069a08e7b7e9e3a7340e8a
  Stored in directory: /root/.cache/pip/wheels/9e/4a/91/e7c3fe95818af06f6db755ee4f86e0c3dfde821606279d8ef8
  Cre

In [ ]:
model = TransformerSummarizer(transformer_type="XLNet",transformer_model_key="xlnet-base-cased")
XLNET_summary = ''.join(model(text, min_length=60))
XLNET_summary

config.json:   0%|          | 0.00/760 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/467M [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/798k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.38M [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


'Scientists say they have discovered a new species of orangutans on Indonesia’s island\nof Sumatra. They say the animals are considered\na new species because of genetic, skeletal and tooth differences. Research into the new species began in 2013, when an orangutan protection group in Sumatra\nfound an injured orangutan in an area far away from the other species. The Tapanuli orangutans also have a different diet and are found only in higher forest\nareas. In addition, the writers of the study are recommending that plans for a hydropower center in\nthe area be stopped by the government. It also recommended that remaining forest in the Sumatran area where the orangutans live be\nprotected.'

#Calculating ROUGE scores for text summaries

In [ ]:
!pip install rouge

In [ ]:
import torch
from rouge import Rouge

# Define the generated summary and the reference summary
generated_summary = "Text summarization with Transformers is efficient for producing precise and relevant summaries."
reference_summary = "Text summarization with Transformers can be used to produce precise and relevant summaries."

# Initialize the ROUGE object
rouge = Rouge()

# Calculate ROUGE for the generated and reference summaries
scores = rouge.get_scores(generated_summary, reference_summary)

# Print the results
print(scores)

[{'rouge-1': {'r': 0.6153846153846154, 'p': 0.6666666666666666, 'f': 0.6399999950080001}, 'rouge-2': {'r': 0.5, 'p': 0.5454545454545454, 'f': 0.5217391254442345}, 'rouge-l': {'r': 0.6153846153846154, 'p': 0.6666666666666666, 'f': 0.6399999950080001}}]


In [ ]:
# Define the generated summary (BERT summary) and the reference summary
generated_summary = '''
Scientists say they have discovered a new species of orangutans on
Indonesia’s island of Sumatra.
They say the animals are considered a new species because of genetic,
skeletal and tooth differences.
Orangutan – which means person of the forest in the Indonesian and
Malay languages - is the world’s biggest tree-living mammal.
Their low numbers make the group the most endangered of all the great
ape species.
He said it puts responsibility on the Indonesian government to help the
species survive.
It also recommended that remaining forest in the Sumatran area where
the orangutans live be protected.
'''
reference_summary = '''
Scientists say they have discovered a new species of orangutans on
Indonesia’s island of Sumatra.
They say the animals are considered a new species because of genetic,
skeletal and tooth differences.
Research into the new species began in 2013, when an orangutan
protection group in Sumatra found an injured orangutan in an area far
away from the other species.
The Tapanuli orangutans also have a different diet and are found only
in higher forest areas.
In addition, the writers of the study are recommending that plans for a
hydropower center in the area be stopped by the government.
It also recommended that remaining forest in the Sumatran area where
the orangutans live be protected.
'''

# Initialize the ROUGE object
rouge = Rouge()
# Calculate ROUGE for the generated and reference summaries
scores = rouge.get_scores(generated_summary, reference_summary)
# Print the results
print(scores)

[{'rouge-1': {'r': 0.5616438356164384, 'p': 0.5694444444444444, 'f': 0.5655172363795482}, 'rouge-2': {'r': 0.4090909090909091, 'p': 0.4891304347826087, 'f': 0.44554454949514755}, 'rouge-l': {'r': 0.5616438356164384, 'p': 0.5694444444444444, 'f': 0.5655172363795482}}]


In [ ]:
# Define the generated summary (GPT summary) and the reference summary
generated_summary = '''
Scientists say they have discovered a new species of orangutans on
Indonesia’s island of Sumatra.
The population differs in several ways from the two existing orangutan
species found in Sumatra and the neighboring island of Borneo.
They say the animals are considered a new species because of genetic,
skeletal and tooth differences.
They live within an area covering about 1,000 square kilometers.
That is because the environment which they depend on is greatly
threatened by development.
In addition, the writers of the study are recommending that plans for a
hydropower center in the area be stopped by the government
'''
reference_summary = '''
Scientists say they have discovered a new species of orangutans on
Indonesia’s island of Sumatra.
They say the animals are considered a new species because of genetic,
skeletal and tooth differences.
Research into the new species began in 2013, when an orangutan
protection group in Sumatra found an injured orangutan in an area far
away from the other species.
The Tapanuli orangutans also have a different diet and are found only
in higher forest areas.
In addition, the writers of the study are recommending that plans for a
hydropower center in the area be stopped by the government.
It also recommended that remaining forest in the Sumatran area where
the orangutans live be protected.
'''
# Initialize the ROUGE object
rouge = Rouge()
# Calculate ROUGE for the generated and reference summaries
scores = rouge.get_scores(generated_summary, reference_summary)
# Print the results
print(scores)

[{'rouge-1': {'r': 0.6438356164383562, 'p': 0.6811594202898551, 'f': 0.661971825989883}, 'rouge-2': {'r': 0.4727272727272727, 'p': 0.5473684210526316, 'f': 0.5073170681975016}, 'rouge-l': {'r': 0.6301369863013698, 'p': 0.6666666666666666, 'f': 0.6478873189476295}}]


##Summary

BERT recall, precision, and f1 are 0.569; 0.569; 0.567 for Rouge-1 and 0.409; 0.489; 0.446 Rouge-2 respectively.
GPT-2 recall, precision, and f1 are 0.64; 0.68; 0.66 for Rouge-1 and 0.47; 0.55 and 0.51 for rouge-2
respectively.